### ISSUE TITLE
NormalizePath is not compatible with mikey179/vfsstream

### ISSUE THREAD

**Issue by zeleznypa** [nette/utils#310]
For testing purposes the package [mikey179/vfsstream](https://packagist.org/packages/mikey179/vfsstream) creates the virtual files with a path starting with `vfs://` but the normalizer removes one `/` character from the protocol part.

https://github.com/nette/utils/blob/d668e02bf5cc786b07abbb5658c97abbb4a00850/src/Utils/FileSystem.php#L283

For example, the: `vfs://directory/file.ext` is the correct "normal" path.

**Comment by dg** [comment]
That's the `schema` `://` `path`, right?

**Comment by zeleznypa** [comment]
Hope it helps: https://github.com/nette/utils/pull/311

**Comment by dg** [comment]
So just use it for the `path` part…

**Comment by zeleznypa** [comment]
Protocol like ssh:// is a valid part of the path. Why to force user to remove the protocol to use the method to add protocol back again?

this method is realy handy, until it os not usable…

**Comment by dg** [comment]
Can you show me a real ssh path?

**Comment by zeleznypa** [comment]
Like here: https://www.arubanetworks.com/techdocs/AOS-CX/10.07/HTML/5200-7851/Content/Chp_Cfg_FW_mgt/FW_mgt_cmds/sft-url.htm

`sftp://user@1.1.1.1/a.txt`

I'm looking for a simple usage like

```php
public function getInstalled(string $vendorPath): array
{
    $installedFilePath = FileSystem::joinPaths($vendorDirectory, 'composer', 'installed.php');
    $installed = require($installedFilePath);
    return $installed;
}
```

When I use a path like `/var/www/project/vendor`, it works like a charm.

But when I want to test it, I'm using the `mikey179/vfsstream`. I have created a virtual folder with a virtual file and then, test the result.

But now it is not working, because the function breaks the path in the protocol part.

There are many [`Supported protocols`](https://www.php.net/manual/en/wrappers.php) that can be used as paths for example in [`file_get_contents`](https://www.php.net/manual/en/function.file-get-contents.php).

Why not allow us to use them?

**Comment by dg** [comment]
And what will the code #311 return for this input `sftp://user@1.1.1.1/../a.txt` ?

**Comment by zeleznypa** [comment]
Almost the same wrong result as before the change:

Before: `sftp:/a.txt`
After: `sftp://a.txt`

**Comment by dg** [comment]
Which is wrong. What about it?

**Comment by zeleznypa** [comment]
Both of them are in this corner case the wrong one

**Comment by dg** [comment]
The current behavior is correct because the function does not support the schema. The schema is not part of the path.  https://github.com/nette/utils/issues/310#issuecomment-2093157317

**Comment by zeleznypa** [comment]
Current behaviour is not correct, because it create total non-valid unusable crap and loose the information about protocol. So you also loose the information, why the result path is wrong

**Comment by zeleznypa** [comment]
Should I burn more time to add support for the server part of the "ssh", "sftp" and some other protocols? Or will you still not accept it?

**Comment by dg** [comment]
Then take a good look at the name of the function. Is it called `normalizeURI()` or `normalizePath()`? Should something called `normalizePath` handle URI normalization? And can you solve such a thing? And do you want to do it?

(please don't do this)

**Comment by zeleznypa** [comment]
It should not. But It also should not break the code.

Pity, it could have been a practical feature.

So please add a big warning into the documentation to not use it anywhere, where the path with protocol can come.

**Comment by dg** [comment]
Yeah, I'll put up a giant red warning. What the path normalization function can't do. It can't normalize query parameters. It can't normalize the schema. It can't validate an email address. It is just piece of crap.

**Comment by dg** [comment]
Btw, so far you haven't shown that you can implement it correctly. How would you handle the normalization so it doesn't break ssh?

**Comment by zeleznypa** [comment]
Please let's stick to professional discussion and leave out the snide remarks.

- Building a file path is a valid use case.
- Concatenating a file path from individual parts is a valid use case.
- Using the path to a physical or virtual folder as the source folder is a valid use case.
- To test the code is a valid use case.

Using the innocent-sounding `joinPaths` function for this purpose appears to be a valid use case, but given the lack of protocol support, it is not.

The warning that the protocol is not supported is a relevant measure, at least as part of the phpDoc above the function.

As in the current solution, where `/../a` remains "denormalized" instead of the correct `/a`, it is valid for many protocols (`ftp`, `sftp`, `ssh` [remote protocols]) to strictly keep the first part.

But that's not what this ticket was about.
This ticket only addresses the part of maintaining support for using VFS for testing purposes.

**Comment by dg** [comment]
![image](https://github.com/nette/utils/assets/194960/7db4f7eb-74d3-4b38-900a-cb77c72ca810)

**Comment by zeleznypa** [comment]
Please look at the https://www.php.net/manual/en/wrappers.php

> Note: The URL syntax used to describe a wrapper only supports the scheme://... syntax. The scheme:/ and scheme: syntaxes are not supported.

the `vfs:/bbb` is not valid, because it will not be interpreted as a `vfs` protocol but as a relative path that begins on the not-valid `vfs:` folder name.

**Comment by zeleznypa** [comment]
Maybe the question is different. Should the `joinPaths` function do the `normqlizePath` also? 

Is it single responsibility?

—-

Sorry for my straight (not so polite) english. It is not my native language. I don’t wont to be or soubds like offensive.

**Comment by Rixafy** [comment]
Flysystem is the most popular PHP library for abstract work with remote files or files in general. There is also a path normalizer https://github.com/thephpleague/flysystem/blob/3.x/src/WhitespacePathNormalizer.php and if the input will be `vfs://directory/file.ext` the output will be `vfs:/directory/file.ext`, so I guess you can have the same argument with `file_get_contents` there, but the quoted text you have attached specifies that is suppots **scheeme**://
.. syntax, that means not only paths.

The point is, `vfs` is a scheme and `://` is what separates scheme and path, almost every normalization function I've seen on the internet does `explode` by `/`, checks if there is `'.'` or `''`, then it skips those things and then does `implode`, that means, `//` won't survive, and it shouldn't, because that's also a reason why do we normalize paths.
https://www.rfc-editor.org/rfc/rfc3986#section-3

> Maybe the question is different. Should the joinPaths function do the normalizePath also?

Then that function would have no meaning, since it's primary function is to remove duplicated `/`.

> Sorry for my straight (not so polite) english. It is not my native language. I don’t wont to be or soubds like offensive.

Don't worry, it's not the english, I've seen some message on slack that wasn't in english and was offensive as hell and I was just hoping that no one replies so you can safely delete it. Next time it's just better ask people to come discuss the topic and not assume you are right and call people to explain another programmer that scheme is part of the path.

I understand it's unexpected, but it's still better than having a condition check for `://` and allowing combination of that 3 characters, because people will start using it for whole URI addresses and then it will only get worse, because in URI, there are usernames, passwords, hosts, ports, parameters..

I don't know how often does this happen, but if it was more common problem, it may help to add check for `://` and throw an user warning that whole URI was sent and it was corrupted by function, but I don't think it's necessary, since you'll probably encounter the problem on the next line when you'll try to read file.

**Comment by dg** [comment]
> Please look at the https://www.php.net/manual/en/wrappers.php
> 
> the `vfs:/bbb` is not valid, because it will not be interpreted as a `vfs` protocol but as a relative path that begins on the not-valid `vfs:` folder name.
> 

`vfs:` is perfectly valid file and folder name.

The only way `normalizePath()` could support URIs is to ignore them. Because there is no general way to normalize URIs. So if the parameter starts with `\w+://`, it will return unchanged. But you still haven't explained why you even want to call `normalizePath` with such a parameter?

**Comment by zeleznypa** [comment]
@Rixafy the argument with the different package look strange, but it’s valid. Thanks for it.

@dg I still think, that the example from the begin is valid, understandable and is still the problem, that should be solved somehow.

Example that shows that “joinPaths” function leads to untestable code.

Warning in the code/documentation, or  “check” for the `://` (as weitten by @Rixafy) will help. At least to show that “it’s not a bug, it’s a feature”

—-

And sorry please if it was sound rude or offensive, I don’t mean it to be so… My apologize

**Comment by Rixafy** [comment]
But as David said, `vfs:` is a valid file and folder name, so if you wanted normalize path `vfs://test.txt`, how could you know that `vfs:` isn't a folder name? So it's only possible to mention it in the docs if it helps.

**Comment by zeleznypa** [comment]
Because the `:` is not a valid character for folder name:

https://stackoverflow.com/q/1976007

**Comment by Rixafy** [comment]
Only on windows, in Unix/Linux only invalid character is `/`.

**Comment by zeleznypa** [comment]
MacOS also don’t let you create such a folder
